[quant][graphmode][fx] qconfig_dict support more types of configurations

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44856 [quant][graphmode][fx] qconfig_dict support more types of configurations**

Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

jerryzh168 committed:
[quant][graphmode][fx] qconfig_dict support more types of configurations

Summary:
Support following format of qconfig_dict
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -123,6 +124,29 @@ def assert_and_get_unique_device(module):
+    a list of tuples of regex and qconfig (module_name_regex)
+    since order is important for regex matching (there might be conflicts and we
+    will take the first match)
+    and is not important for other qconfig types.
```
thoughts about just making all of them match in an ordered fashion?  It could be simpler both to the user as well as in the implementation as everything will work the same way.

jerryzh168 replied:
looks like there is a cost to do that, which is making the qconfig lookup O(N) instead of O(1), where N is the number of entries of qconfig_dict. I feel this is an implementation detail, it's good enough for user to know the precedence between different types of configuration, and the matching within one type of configuration can still be considered as matched in order (unless user provides duplicate keys, we may check for this if needed).

vkuzo replied:
I'd be surprised if the performance of this block of code was a bottleneck.  For qconfig sizes we can reasonably expect to see (let's say a median size of 1 rule, and maybe p95 of 100 rules (just a guess)), would this matter?

jerryzh168 replied:
I don't think this will be a bottleneck, but it doesn't sound like a worthwhile trade off to me. since
1). it looks more natural to do dictionary lookup for the qconfigs where order does not matter
2). it does not affect user experience since this is an implementation detail
3). we are reusing propagate_qconfig_ from eager mode, which expects a dictionary, will need to update the matching there as well if we want to remove the flatten step.

jerryzh168 replied:
actually I think an alternative is convert them to ordered dict, and we can probably use the same data format for different qconfig types, although still different code to get the qconfig. does that sounds good?

vkuzo replied:
> 2). it does not affect user experience since this is an implementation detail

sg.  I don't feel too strongly

> actually I think an alternative is to convert them to ordered dict

in the API, or in the implementation?  In implementation would make sense.  For the API, that might make it more cumbersome for users.

jerryzh168 replied:
> in the API, or in the implementation? In implementation would make sense. For the API, that might make it more cumbersome for users.

in implementation.

vkuzo reviewed (commented):
awesome!

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -332,7 +333,8 @@ def forward(self, x):
         m = symbolic_trace(m)
-        qconfig_dict = {'': default_qconfig, 'conv2': None}
+        qconfig_dict = {"": default_qconfig,
+                        "module_name": [("conv2", None)]}
```
Is there a reason to prefer ("conv2", None) to  "conv2": None?  Wouldnt a dict be more natural here?

jerryzh168 replied:
yeah, i agree. the reason we use ("conv2", None) is because "module_name_regex" needs to be matched in order and we want to keep the type for all configurations consistent.

raghuramank100 commented on torch/quantization/fx/quantize.py:
```diff
@@ -168,23 +192,74 @@ def __init__(self):
-    def _generate_qconfig_map(self, root, input_graph):
-        def get_qconfig(module):
-            return module.qconfig if hasattr(module, 'qconfig') else None
+    def _generate_qconfig_map(self,
```
If we go with a dict for convert, this logic can be shared with the convert_dict to control actual quantization.

jerryzh168 replied:
we still convert that to dict internally. only 'module_name_regex' is left at list of tuples since order is required for that.

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -168,23 +192,74 @@ def __init__(self):
+    def _generate_qconfig_map(self,
+                              root,
+                              input_graph,
+                              flattened_qconfig_dict,
+                              module_name_regex_qconfig_list):
```
any chance we can annotate these types in the docblock, to help with readability?

jerryzh168 replied:
changed to using OrderedDict, hopefully it's clearer now

vkuzo commented on torch/quantization/quantize_fx.py:
```diff
@@ -115,16 +115,34 @@ def quantize_static_fx(model, qconfig_dict, run_fn, run_args, inplace=False, deb
+    # optional, used for module and function types
+    # could also be split into module_types and function_types if we prefer
+    "object_type": [
+      (nn.Conv2d, qconfig?),
+      (F.add, qconfig?),
```
sorry, didn't know these would make it to the code :) maybe we can use `torch.nn` and `torch.nn.functional`?

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] qconfig_dict support more types of configurations"


Summary:
Support following qconfig_dict format:
```python
qconfig_dict = {
    # optional, global config
    "": qconfig?,

    # optional, used for module and function types
    # could also be split into module_types and function_types if we prefer
    "object_type": [
      (nn.Conv2d, qconfig?),
      (F.add, qconfig?),
      ...,
    ],

    # optional, used for module names
    "module_name": [
      ("foo.bar", qconfig?)
      ...,
    ],

    # optional, matched in order, first match takes precedence
    "module_name_regex": [
      ("foo.*bar.*conv[0-9]+", qconfig?)
      ...,
    ]
    # priority (in increasing order): global, object_type, module_name_regex, module_name
    # qconfig == None means fusion and quantization should be skipped for anything
    # matching the rule
}
```
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23751304](https://our.internmc.facebook.com/intern/diff/D23751304)

[ghstack-poisoned]
